src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetriesBehavior.cs

```diff
@@ -30,39 +43,51 @@ public SecondLevelRetriesBehavior(SecondLevelRetryPolicy retryPolicy, string loc
             }
             catch (Exception ex)
             {
-                var message = context.Message;
-                var currentRetry = GetNumberOfRetries(message.Headers) + 1;
+                failureInfoStorage.MarkForDeferralForSecondLevelRetry(context.Message.MessageId, ExceptionDispatchInfo.Capture(ex));
```

Comment:
Currently when exception is thrown and `SLR` has reached it's limit we still rollback the message and throw in `DeferMessageForSecondLevelRetry` on next receive. We could bubble the exception here but would need to check the policy as well.